pkg/cloudprovider/providers/gce/gce_disks.go

```diff
@@ -867,63 +945,73 @@ func (gce *GCECloud) GetAutoLabelsForPD(name string, zone string) (map[string]st
 	return labels, nil
 }
 
-// Returns a GCEDisk for the disk, if it is found in the specified zone.
-// If not found, returns (nil, nil)
-func (gce *GCECloud) findDiskByName(diskName string, zone string) (*GCEDisk, error) {
+// getDiskByName returns a GCEDisk for the disk, if it is found in the specified zone.
+// Otherwise returns a DiskNotFoundError.
+func (gce *GCECloud) getDiskByName(diskName string, zone string) (*GCEDisk, error) {
 	mc := newDiskMetricContextZonal("get", gce.region, zone)
 	disk, err := gce.manager.GetDiskFromCloudProvider(zone, diskName)
-	if err == nil {
-		return disk, mc.Observe(nil)
-	}
-	if !isHTTPErrorCode(err, http.StatusNotFound) {
-		return nil, mc.Observe(err)
+	mc.Observe(err)
+
+	if isHTTPErrorCode(err, http.StatusNotFound) {
+		return nil, &DiskNotFoundError{regional: false, diskName: diskName, diskZone: zone}
 	}
-	return nil, mc.Observe(nil)
-}
 
-// Like findDiskByName, but returns an error if the disk is not found
-func (gce *GCECloud) getDiskByName(diskName string, zone string) (*GCEDisk, error) {
-	disk, err := gce.findDiskByName(diskName, zone)
-	if disk == nil && err == nil {
-		return nil, fmt.Errorf("GCE persistent disk not found: diskName=%q zone=%q", diskName, zone)
+	if err != nil {
+		return nil, err
 	}
-	return disk, err
+	return disk, nil
 }
 
-// Returns a GCEDisk for the regional disk, if it is found.
-// If not found, returns (nil, nil)
-func (gce *GCECloud) findRegionalDiskByName(diskName string) (*GCEDisk, error) {
+// getRegionalDiskByName returns a GCEDisk for the regional disk, if it is found.
+// Otherwise returns a DiskNotFoundError.
+func (gce *GCECloud) getRegionalDiskByName(diskName string) (*GCEDisk, error) {
 	mc := newDiskMetricContextRegional("get", gce.region)
 	disk, err := gce.manager.GetRegionalDiskFromCloudProvider(diskName)
-	if err == nil {
-		return disk, mc.Observe(nil)
+	mc.Observe(err)
+
+	if isHTTPErrorCode(err, http.StatusNotFound) {
+		return nil, &DiskNotFoundError{regional: true, diskName: diskName}
 	}
-	if !isHTTPErrorCode(err, http.StatusNotFound) {
-		return nil, mc.Observe(err)
+
+	if err != nil {
+		return nil, err
 	}
-	return nil, mc.Observe(nil)
+	return disk, nil
```

Comment:
Double check this behavior change is intended.

Reply:
I changed it in order to add the new DiskNotFoundError, but overlooked the fact that StatusNotFound errors shouldn't be observed. Thanks!